src/v/net/conn_quota.cc

```diff
@@ -0,0 +1,440 @@
+// Copyright 2022 Vectorized, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "conn_quota.h"
+
+#include "config/configuration.h"
+#include "hashing/xx.h"
+#include "rpc/logger.h"
+#include "seastar/core/coroutine.hh"
+#include "ssx/future-util.h"
+#include "vassert.h"
+#include "vlog.h"
+
+namespace net {
+
+conn_quota::units::~units() {
+    if (_live) {
+        (*_quotas).get().put(_addr);
+    }
+}
+
+conn_quota::conn_quota(conn_quota::config_fn cfg_f) noexcept
+  : _cfg(cfg_f()) {
+    if (ss::this_shard_id() == total_shard) {
+        if (_cfg.max_connections()) {
+            // Initialize state for enforcement of global connection count
+            total_home.max = _cfg.max_connections().value();
+            total_home.available = total_home.max;
+        }
+    }
+
+    _cfg.max_connections.watch([this]() {
+        if (ss::this_shard_id() == total_shard) {
+            if (_cfg.max_connections()) {
+                vlog(
+                  rpc::rpclog.info,
+                  "Connection count limit updated to {}",
+                  _cfg.max_connections().value());
+                update_limit({}, total_home, _cfg.max_connections().value());
+            } else {
+                total_home = home_allowance{};
+                vlog(rpc::rpclog.info, "Connection count limit disabled");
+            }
+        } else {
+            if (!_cfg.max_connections()) {
+                // Reset, in anticipation of limit being re-enabled later
+                // (do not want to keep borrowed units around)
+                total_remote = remote_allowance{};
+            }
+        }
+    });
+
+    _cfg.max_connections_per_ip.watch([this]() {
+        if (!_cfg.max_connections_per_ip()) {
+            if (ss::this_shard_id() == 0) {
+                vlog(
+                  rpc::rpclog.info, "Connection count per-IP limit disabled");
+            }
+            ip_home.clear();
+            ip_remote.clear();
```

Comment:
I ended up switching to wrapping all the `home_allowance` instances in lw_shared_ptr to deal with this.  There wasn't much of an alternative, to keep the config updates synchronous.  Switched the enclosing structures to flat_hash_map so that we aren't taking the allocation overhead of the node_hash_map as well as the lw_shared_ptr.